feat(header): add Header component

nathanjamal opened the pull request:
## Pull Request

### Description

* feat(header): add Header component

created a Header component. Incomplete at the moment and lacking proper documentation

HEYUI-283

* feat(header): control lang select menu with keyboard

allow language select menu to be opened and closed with keyboard + added some aria data

HEYUI-283

* feat(header): added keyboard support

Added keyboard support, Aria labels and some tests for the hooks and utils

HEYUI-283

* feat(header): added navigation component

added navigation component that builds a nav and subnav based on an object

HEYUI-283

* feat(header): burger menu functionality

added burger menu functionality and moved some functions out of view files

HEYUI-283

* feat(header): mobile search open and close functionality

added toggle on and off for mobile search

HEYUI-283

* docs(header): move header into organisms folder

Merge main and put header under organisms in the storybook

HEYUI-283

* feat(header): move header items that are not visible at mobile into the burger menu

language select and account are not visible in the header in mobile breakpoints. so we move them
into the nav menu

HEYUI-283

* change to single semi-colon syntax

* eslint disable in a test

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change

### How do I test this

- Add steps to test
- in bullet point format
- preferably you can add link to the storybook build in the PR

## Checklist

### Did you remember to take care of the following?

- [ ] `npm i` – for new NPM dependencies.
- [ ] `npm run lint` - to check for linting issues
- [ ] `npm run test` - to run unit tests
- [ ] `npm run test:sh:docker` - to run screenshot tests, [detail instruction](https://hey-car.github.io/heycar-uikit/main/?path=/docs/guidelines-screenshot-testing--page)

### New Feature / Bug Fix

- [ ] Run unit tests to ensure all existing tests are still passing.
- [ ] Add new passing unit tests to cover the code introduced by your pr.

Thanks for contributing!

nathanjamal committed:
feat(header): add Header component (#185)

* feat(header): add Header component

created a Header component. Incomplete at the moment and lacking proper documentation

HEYUI-283

* feat(header): control lang select menu with keyboard

allow language select menu to be opened and closed with keyboard + added some aria data

HEYUI-283

* feat(header): added keyboard support

Added keyboard support, Aria labels and some tests for the hooks and utils

HEYUI-283

* feat(header): added navigation component

added navigation component that builds a nav and subnav based on an object

HEYUI-283

* feat(header): burger menu functionality

added burger menu functionality and moved some functions out of view files

HEYUI-283

* feat(header): mobile search open and close functionality

added toggle on and off for mobile search

HEYUI-283

* docs(header): move header into organisms folder

Merge main and put header under organisms in the storybook

HEYUI-283

* feat(header): move header items that are not visible at mobile into the burger menu

language select and account are not visible in the header in mobile breakpoints. so we move them
into the nav menu

HEYUI-283

* change to single semi-colon syntax

* eslint disable in a test

nathanjamal committed:
test(header): added unit and snapshot tests to header  (#187)

* feat(header): add Header component

created a Header component. Incomplete at the moment and lacking proper documentation

HEYUI-283

* feat(header): control lang select menu with keyboard

allow language select menu to be opened and closed with keyboard + added some aria data

HEYUI-283

* feat(header): added keyboard support

Added keyboard support, Aria labels and some tests for the hooks and utils

HEYUI-283

* feat(header): added navigation component

added navigation component that builds a nav and subnav based on an object

HEYUI-283

* feat(header): burger menu functionality

added burger menu functionality and moved some functions out of view files

HEYUI-283

* feat(header): mobile search open and close functionality

added toggle on and off for mobile search

HEYUI-283

* docs(header): move header into organisms folder

Merge main and put header under organisms in the storybook

HEYUI-283

* feat(header): move header items that are not visible at mobile into the burger menu

language select and account are not visible in the header in mobile breakpoints. so we move them
into the nav menu

HEYUI-283

* test(header): added unit and snapshot tests to header

added unit test for all hooks and util functions plus for header component. andded a single snapshot
test for header.

HEYUI-304

* test(header): unit tests for some rendering and interaction details

unit tests for some rendering and interaction details

HEYUI-304

nathanjamal committed:
feat(header): added documentation and some initial screenshot tests (#189)

* feat(header): added documentation and some initial screenshot tests

set up the storybook canvas, added some documentation and set up some basic screenshot tests. more
tests to come

HEYUI-305

* get nav data from nav mock file

nathanjamal commented on .storybook/scope.ts:
```diff
@@ -33,4 +34,5 @@ export default {
   ...dateUtils,
   ...knobs,
   ...AllIcons,
+  ...headerMocks,
```
needed to get the header mock/default data into the docs

nathanjamal commented on package.json:
```diff
@@ -84,16 +84,16 @@
-    "@storybook/addon-knobs": "^6.4.0",
-    "@storybook/addon-viewport": "^6.5.8",
-    "@storybook/addons": "^6.4.19",
+    "@storybook/addon-controls": "~6.4.20",
```
storybook 6.5.x introduces issues with persisting the storybook controls in the url, which is something we rely on for screenshot testing

nathanjamal commented on packages/header/src/__tests__/Header.mock.tsx:
```diff
@@ -0,0 +1,63 @@
+    email: {
+      value: 'support@heycar.com',
+    },
+    // app: {
```
app badges are coming in a subsequent PR. an App badge Component needs to be built first

nathanjamal commented on packages/pagination/src/__tests__/hooks/useBreakpoint.hook.test.ts:
```diff
@@ -1,3 +1,4 @@
+//NOTE: This has been moved to theme package. Please update references and remove this version when possible
```
once this is merged we can create a ticket to remove this package specific version of useBreakpoint

nathanjamal committed:
test(header): update tests and readme

update some failing tests and proper info in the readme

HEYUI-302

nathanjamal committed:
atempt to fix rollup error

Nazzanuk commented on packages/themes/src/hooks/useBreakpoint.hook.ts:
```diff
@@ -0,0 +1,102 @@
+import { useEffect, useState } from 'react';
```
late in the day but is it possible to remove this hook entirely? 

I thought we agreed something similar in the UK - it doesn't run server side so I'm guessing (haven't tested) that initially user will get mobile view before it adjusts as the css will render before the js runs.

Would it be possible to just have 2 dedicated views, one for mobile/tablet and one for desktop and we just show and hide each one based only on css? (this would result in a bigger DOM but less reliance on js)

nathanjamal replied:
@Nazzanuk If that's the move, then that's the move 🫡

nathanjamal committed:
feat(header): swap js breakpoint detection for seperate views

removed the js breakpoint detection and refactor to have seperate views for mobile and desktop that
are show via CSS

HEYUI-302

nathanjamal committed:
feat(header): display favorites count

now displays faves count up to 99. updated tests to reflect

HEYUI-302

nathanjamal committed:
style updates

nathanjamal committed:
removed unused import

nathanjamal committed:
update snapshot

esraltintas commented on packages/header/src/Header.tsx:
```diff
@@ -0,0 +1,291 @@
+      favoritesItemConfig,
+      langItemConfig,
+      LinkComponent,
+      locale = DEFAULT_LOCALE,
```
Hey @nathanjamal could you check test coverage for here, please?

esraltintas commented:
Good job @nathanjamal could you check the test coverage issues on header.tsx please?

nathanjamal committed:
more tests

nathanjamal committed:
Merge remote-tracking branch 'origin/main' into HEYUI-302

nathanjamal committed:
remove eslint disable comment

nathanjamal committed:
remove unneeded changes

nathanjamal commented on packages/header/src/Header.tsx:
```diff
@@ -0,0 +1,291 @@
+                    aria-haspopup="menu"
+                    aria-label={`${locale.langListHeading} - ${locale.spaceBarNotification}`}
+                    className={`${styles.horizontalNavOnly} ${styles.item} ${
+                      isLangListOpen ? styles.focused : ''
```
this is covered in screenshot tests

nathanjamal commented on packages/header/src/Header.tsx:
```diff
@@ -0,0 +1,291 @@
+                  )
+                }
+              >
+                {isNavTrayOpen ? <Close /> : <Hamburger />}
```
covered in screen shot tests

nathanjamal commented on packages/header/src/Header.tsx:
```diff
@@ -0,0 +1,291 @@
+         *** Navigation
+         */}
+        <div
+          className={`${styles.navWrapper} ${isNavTrayOpen ? 'navOpen' : ''}`}
```
covered in screen shot tests

nathanjamal committed:
more tests

nathanjamal committed:
remove cn dependancy

nathanjamal committed:
remove cn dependancy again

nathanjamal committed:
more cn removal

nathanjamal committed:
update classnames version

nathanjamal committed:
update classnames version

nathanjamal committed:
Merge remote-tracking branch 'origin/main' into HEYUI-302

nathanjamal committed:
try to fix classnames default export error
